### ISSUE TITLE
Some (newer?) URL's on Banbye.com don't work any more

### ISSUE THREAD

**Issue by nicolaasjan** [yt-dlp/yt-dlp#8584]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

The Netherlands

### Provide a description that is worded well enough to be understood

The URL:
https://banbye.com/watch/v_a_gPFuC9LoW5
gives `HTTP Error 500: Internal Server Error.`
(see the complete verbose output)

However giving yt-dlp the `index.m3u8` URL succeeds:

```shell
./yt-dlp -v --ignore-config -N 6 "https://tc2.banbye.com:31001/live/hls/dl9hX2dQRnVDOUxvVzU_480/index.m3u8"
[debug] Command-line config: ['-v', '--ignore-config', '-N', '6', 'https://tc2.banbye.com:31001/live/hls/dl9hX2dQRnVDOUxvVzU_480/index.m3u8']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.11.14 from yt-dlp/yt-dlp [a9d3f4b20] (zip)
[debug] Python 3.8.10 (CPython x86_64 64bit) - Linux-5.4.0-166-generic-x86_64-with-glibc2.29 (OpenSSL 1.1.1f  31 Mar 2020, glibc 2.31)
[debug] exe versions: ffmpeg N-112741-gcd7b352c53-20231112 (setts), ffprobe N-112741-gcd7b352c53-20231112, phantomjs 2.1.1, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.19.0, brotli-1.1.0, certifi-2023.07.22, mutagen-1.47.0, requests-2.31.0, secretstorage-3.3.3, sqlite3-3.31.1, urllib3-2.0.6, websockets-11.0.3, xattr-0.10.1
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests
[debug] Loaded 1899 extractors
[generic] Extracting URL: https://tc2.banbye.com:31001/live/hls/dl9hX2dQRnVDOUxvVzU_480/index.m3u8
[generic] index: Downloading webpage
[debug] Identified a direct video link
[generic] index: Downloading m3u8 information
[generic] index: Checking m3u8 live status
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] index: Downloading 1 format(s): 0
[debug] Invoking hlsnative downloader on "https://tc2.banbye.com:31001/live/hls/dl9hX2dQRnVDOUxvVzU_480/index.m3u8"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 491
[download] Destination: index [index].mp4
[download] 100% of  341.26MiB in 00:00:50 at 6.76MiB/s
[debug] ffprobe command line: ffprobe -hide_banner -show_format -show_streams -print_format json 'file:index [index].mp4'
[debug] ffmpeg command line: ffprobe -show_streams 'file:index [index].mp4'
[FixupM3u8] Fixing MPEG-TS in MP4 container of "index [index].mp4"
[debug] ffmpeg command line: ffmpeg -y -loglevel repeat+info -i 'file:index [index].mp4' -map 0 -dn -ignore_unknown -c copy -f mp4 -bsf:a aac_adtstoasc -movflags +faststart 'file:index [index].temp.mp4'

```
Example of a random other video for which this issue does **not** occur:

```shell
./yt-dlp -N 6 -v --ignore-config "https://banbye.com/watch/v_PUyyINwXAt94"
[debug] Command-line config: ['-N', '6', '-v', '--ignore-config', 'https://banbye.com/watch/v_PUyyINwXAt94']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.11.14 from yt-dlp/yt-dlp [a9d3f4b20] (zip)
[debug] Python 3.8.10 (CPython x86_64 64bit) - Linux-5.4.0-166-generic-x86_64-with-glibc2.29 (OpenSSL 1.1.1f  31 Mar 2020, glibc 2.31)
[debug] exe versions: ffmpeg N-112741-gcd7b352c53-20231112 (setts), ffprobe N-112741-gcd7b352c53-20231112, phantomjs 2.1.1, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.19.0, brotli-1.1.0, certifi-2023.07.22, mutagen-1.47.0, requests-2.31.0, secretstorage-3.3.3, sqlite3-3.31.1, urllib3-2.0.6, websockets-11.0.3, xattr-0.10.1
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests
[debug] Loaded 1899 extractors
[BanBye] Extracting URL: https://banbye.com/watch/v_PUyyINwXAt94
[BanBye] v_PUyyINwXAt94: Downloading JSON metadata
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] v_PUyyINwXAt94: Downloading 1 format(s): http-480p
[debug] Invoking http downloader on "https://cdn.banbye.com/video/v_PUyyINwXAt94/480.mp4"
[download] Destination: Szokujący pokaz hipokryzji Morawieckiego i Dudy？! Ambasadorze Izraela： ZAMILCZ! NIE JESTEŚ U SIEBIE! Vlog Marcina Roli [v_PUyyINwXAt94].mp4
[download] 100% of   76.20MiB in 00:00:03 at 19.35MiB/s
```

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
./yt-dlp -vU --ignore-config https://banbye.com/watch/v_a_gPFuC9LoW5
[debug] Command-line config: ['-vU', '--ignore-config', 'https://banbye.com/watch/v_a_gPFuC9LoW5']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.11.14 from yt-dlp/yt-dlp [a9d3f4b20] (zip)
[debug] Python 3.8.10 (CPython x86_64 64bit) - Linux-5.4.0-166-generic-x86_64-with-glibc2.29 (OpenSSL 1.1.1f  31 Mar 2020, glibc 2.31)
[debug] exe versions: ffmpeg N-112741-gcd7b352c53-20231112 (setts), ffprobe N-112741-gcd7b352c53-20231112, phantomjs 2.1.1, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.19.0, brotli-1.1.0, certifi-2023.07.22, mutagen-1.47.0, requests-2.31.0, secretstorage-3.3.3, sqlite3-3.31.1, urllib3-2.0.6, websockets-11.0.3, xattr-0.10.1
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests
[debug] Loaded 1899 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: stable@2023.11.14 from yt-dlp/yt-dlp
yt-dlp is up to date (stable@2023.11.14 from yt-dlp/yt-dlp)
[BanBye] Extracting URL: https://banbye.com/watch/v_a_gPFuC9LoW5
[BanBye] v_a_gPFuC9LoW5: Downloading JSON metadata
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] v_a_gPFuC9LoW5: Downloading 1 format(s): http-480p
[debug] Invoking http downloader on "https://cdn.banbye.com/video/v_a_gPFuC9LoW5/480.mp4"
[download] Got error: HTTP Error 500: Internal Server Error. Retrying (1/10)...
[download] Got error: HTTP Error 500: Internal Server Error. Retrying (2/10)...
[download] Got error: HTTP Error 500: Internal Server Error. Retrying (3/10)...
[download] Got error: HTTP Error 500: Internal Server Error. Retrying (4/10)...
[download] Got error: HTTP Error 500: Internal Server Error. Retrying (5/10)...
[download] Got error: HTTP Error 500: Internal Server Error. Retrying (6/10)...
[download] Got error: HTTP Error 500: Internal Server Error. Retrying (7/10)...
[download] Got error: HTTP Error 500: Internal Server Error. Retrying (8/10)...
[download] Got error: HTTP Error 500: Internal Server Error. Retrying (9/10)...
[download] Got error: HTTP Error 500: Internal Server Error. Retrying (10/10)...
[download] Got error: HTTP Error 500: Internal Server Error. Giving up after 10 retries
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "./yt-dlp/__main__.py", line 17, in <module>
    yt_dlp.main()
  File "./yt-dlp/yt_dlp/__init__.py", line 1008, in main
    _exit(*variadic(_real_main(argv)))
  File "./yt-dlp/yt_dlp/__init__.py", line 998, in _real_main
    return ydl.download(all_urls)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 3511, in download
    self.__download_wrapper(self.extract_info)(
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 3486, in wrapper
    res = func(*args, **kwargs)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1556, in extract_info
    return self.__extract_info(url, self.get_info_extractor(key), download, extra_info, process)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1567, in wrapper
    return func(self, *args, **kwargs)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1723, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1782, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 2922, in process_video_result
    self.process_info(new_info)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 3388, in process_info
    success, real_download = self.dl(temp_filename, info_dict)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 3109, in dl
    return fd.download(name, new_info, subtitle)
  File "./yt-dlp/yt_dlp/downloader/common.py", line 455, in download
    ret = self.real_download(filename, info_dict)
  File "./yt-dlp/yt_dlp/downloader/http.py", line 362, in real_download
    for retry in RetryManager(self.params.get('retries'), self.report_retry):
  File "./yt-dlp/yt_dlp/utils/_utils.py", line 5114, in __iter__
  File "./yt-dlp/yt_dlp/downloader/common.py", line 401, in report_retry
    RetryManager.report_retry(
  File "./yt-dlp/yt_dlp/utils/_utils.py", line 5121, in report_retry
  File "./yt-dlp/yt_dlp/downloader/common.py", line 404, in <lambda>
    error=IDENTITY if not fatal else lambda e: self.report_error(f'\r[download] Got error: {e}'),
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1045, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 973, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())
```

**Comment by PatrykMis** [comment]
@seproDev @bashonly @Grub4K I can confirm this problem occurs with all new videos not included in yt-dlp's test set.

What I discovered is that the browser tries to fetch the m3u8 playlist URL via `{self._API_BASE}/videos/{video_id}/url`. The issue is that for the API to return a JSON with a valid URL, a specially crafted `POST` request is required, which includes at least the following fields:

* `X-Banbye-Tuid` generated dynamically
* `'User-Agent': 'Mozilla/5.0 (Linux; Android {random_version}; Nexus 5 Build/{random_string}) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/{current_version_used} Mobile Safari/537.36'`
* `Referer: https://banbye.com/`
* Other fields such as:

```http
Sec-Ch-Ua-Mobile: ?1
Sec-Ch-Ua-Platform: "Android"
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-site
Sec-GPC: 1
```

Regardless of whether mobile or desktop is used, the header fields are similar.

If the POST request is formed properly, we can receive a JSON object like this in case of the URL from the first comment (formatted for better readability):

```json
{
    "url": "https://tc2l3.banbye.com",
    "rid": "vw_Rjj4CxVZS76tkFdpeKgFo",
    "src": {
        "hls": {
            "masterPlaylist": "https://tc2l3.banbye.com/edge/hls/dl9hX2dQRnVDOUxvVzU.m3u8",
            "levels": {
                "144": "https://tc2l3.banbye.com/edge/hls/dl9hX2dQRnVDOUxvVzU_144/index.m3u8",
                "480": "https://tc2l3.banbye.com/edge/hls/dl9hX2dQRnVDOUxvVzU_480/index.m3u8"
            }
        }
    }
}
```

If it's not proper, the API returns `http/204` and 'not found' or similar message.

In conclusion, it seems the m3u8 playlists are now obtained with a prior cookie set, similar to YouTube.

**Comment by seproDev** [comment]
Please do not mass ping people.

The `X-Banbye-Tuid` has nothing to do with this. The problem is that the extractor constructs the URL itself, and for newer videos the site has switched to using HLS.
I think something like this should work:
```diff
diff --git a/yt_dlp/extractor/banbye.py b/yt_dlp/extractor/banbye.py
index d10bdf8da..d0dbb4936 100644
--- a/yt_dlp/extractor/banbye.py
+++ b/yt_dlp/extractor/banbye.py
@@ -4,9 +4,12 @@
 from .common import InfoExtractor
 from ..utils import (
     InAdvancePagedList,
+    determine_ext,
     format_field,
+    int_or_none,
     traverse_obj,
     unified_timestamp,
+    url_or_none,
 )
 
 
@@ -91,11 +94,23 @@ def _real_extract(self, url):
             'id': f'{quality}p',
             'url': f'{self._CDN_BASE}/video/{video_id}/{quality}.webp',
         } for quality in [48, 96, 144, 240, 512, 1080]]
-        formats = [{
-            'format_id': f'http-{quality}p',
-            'quality': quality,
-            'url': f'{self._CDN_BASE}/video/{video_id}/{quality}.mp4',
-        } for quality in data['quality']]
+
+        formats = []
+        url_data = self._download_json(f'{self._API_BASE}/videos/{video_id}/url', video_id, data=b'')
+        for format_id, url in traverse_obj(url_data, (
+                'src', ('mp4', 'hls'), 'levels', {dict.items}, lambda _, v: url_or_none(v[1]))):
+            if determine_ext(url) == 'm3u8':
+                fmts = self._extract_m3u8_formats(url, video_id, 'mp4', m3u8_id='hls', fatal=False)
+                for f in fmts:
+                    f['height'] = int_or_none(format_id)
+                formats.extend(fmts)
+            else:
+                formats.append({
+                    'url': url,
+                    'format_id': format_id,
+                    'height': int_or_none(format_id),
+                })
 
         return {
             'id': video_id,
```

**Comment by PatrykMis** [comment]
You're right. It started to work for some more videos, but I feel the more reliable method is to get formats from master playlist, not from JSON. For the latest videos, like: `https://banbye.com/watch/v_B0rsKWsr-aaa`, the following json object is returned:

```json
{
    "url": "https://tc2l3.banbye.com",
    "rid": "vw_AJKoJEpXz296UsgVWwDLB",
    "videoSrc": "video/v_B0rsKWsr-aaa/v/index.m3u8",
    "src": {
        "hls": {
            "masterPlaylist": "https://tc2l3.banbye.com/edge/video/v_B0rsKWsr-aaa/v/index.m3u8",
            "levels": {}
        }
    }
}
```

so, yt-dlp fails for obvious reasons.
